File: backend/services/github_service.py
import httpx
import logging
from typing import List, Dict, Optional
from fastapi import HTTPException

logger = logging.getLogger(__name__)


class GitHubService:
    """Service for interacting with GitHub API."""
    
    def __init__(self):
        self.base_url = "https://api.github.com"
    
    async def fetch_maintainer_repos(
        self, 
        github_access_token: str, 
        existing_repos: List[str]
    ) -> List[Dict[str, str]]:
        """
        Fetch all repositories where the user has admin, maintain, or push permissions,
        excluding repos already added to the website.
        
        Args:
            github_access_token: GitHub OAuth token
            existing_repos: List of repo full names already added
            
        Returns:
            List of dicts with 'owner' and 'repo' keys
        """
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/user/repos",
                    params={"per_page": 100, "type": "owner,collaborator"},
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json"
                    },
                    timeout=30.0
                )
                
                if response.status_code != 200:
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"GitHub API error: {response.status_code} {response.text}"
                    )
                
                repos = response.json()
                
                # Filter repos with required permissions and not in existing_repos
                filtered_repos = []
                for repo in repos:
                    permissions = repo.get('permissions', {})
                    has_permission = (
                        permissions.get('admin') or 
                        permissions.get('maintain') or 
                        permissions.get('push')
                    )
                    
                    if has_permission and repo['full_name'] not in existing_repos:
                        filtered_repos.append({
                            'owner': repo['owner']['login'],
                            'repo': repo['name']
                        })
                
                return filtered_repos
                
        except httpx.TimeoutException:
            raise HTTPException(status_code=504, detail="GitHub API request timed out")
        except Exception as e:
            logger.error(f"Failed to fetch maintainer repos: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to fetch maintainer repos: {str(e)}")
    
    async def fetch_unmerged_pull_requests(
        self, 
        github_access_token: str, 
        owner: str, 
        repo: str
    ) -> List[Dict[str, any]]:
        """
        Fetch only pull requests that are open and not merged.
        
        Args:
            github_access_token: GitHub OAuth token
            owner: Repository owner
            repo: Repository name
            
        Returns:
            List of dicts with 'number' and 'title' keys
        """
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}/pulls",
                    params={"state": "open"},
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json"
                    },
                    timeout=30.0
                )
                
                if response.status_code != 200:
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"GitHub API error: {response.status_code} {response.text}"
                    )
                
                pulls = response.json()
                
                # Filter only unmerged PRs (merged_at is null)
                unmerged_prs = [
                    {
                        'number': pr['number'],
                        'title': pr['title']
                    }
                    for pr in pulls
                    if pr.get('merged_at') is None
                ]
                
                return unmerged_prs
                
        except httpx.TimeoutException:
            raise HTTPException(status_code=504, detail="GitHub API request timed out")
        except Exception as e:
            logger.error(f"Failed to fetch unmerged PRs for {owner}/{repo}: {e}")
            raise HTTPException(
                status_code=500, 
                detail=f"Failed to fetch unmerged PRs for {owner}/{repo}: {str(e)}"
            )
    
    async def comment_on_pull_request(
        self,
        github_access_token: str,
        owner: str,
        repo: str,
        pr_number: int,
        comment_text: str
    ) -> Dict:
        """
        Post a comment on a pull request after verifying repo access.
        
        Args:
            github_access_token: GitHub OAuth token
            owner: Repository owner
            repo: Repository name
            pr_number: Pull request number
            comment_text: Comment body text
            
        Returns:
            GitHub API response for the created comment
        """
        try:
            async with httpx.AsyncClient() as client:
                # First, verify repo access
                repo_response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}",
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json"
                    },
                    timeout=30.0
                )
                
                if repo_response.status_code != 200:
                    raise HTTPException(
                        status_code=repo_response.status_code,
                        detail=f"Cannot access repository: {repo_response.status_code} {repo_response.text}"
                    )
                
                repo_data = repo_response.json()
                permissions = repo_data.get('permissions', {})
                
                # Check if user has required permissions
                if not (permissions.get('admin') or permissions.get('maintain') or permissions.get('push')):
                    raise HTTPException(
                        status_code=403,
                        detail="Insufficient permissions to comment on this repository"
                    )
                
                # Post the comment
                comment_response = await client.post(
                    f"{self.base_url}/repos/{owner}/{repo}/issues/{pr_number}/comments",
                    json={"body": comment_text},
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json",
                        "Content-Type": "application/json"
                    },
                    timeout=30.0
                )
                
                if comment_response.status_code not in [200, 201]:
                    raise HTTPException(
                        status_code=comment_response.status_code,
                        detail=f"Failed to post comment: {comment_response.status_code} {comment_response.text}"
                    )
                
                return comment_response.json()
                
        except HTTPException:
            raise
        except httpx.TimeoutException:
            raise HTTPException(status_code=504, detail="GitHub API request timed out")
        except Exception as e:
            logger.error(f"Failed to comment on PR #{pr_number} in {owner}/{repo}: {e}")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to comment on PR #{pr_number} in {owner}/{repo}: {str(e)}"
            )
    
    async def comment_on_issue(
        self,
        github_access_token: str,
        owner: str,
        repo: str,
        issue_number: int,
        comment_text: str
    ) -> Dict:
        """
        Post a comment on a GitHub issue after verifying repo access.
        
        Args:
            github_access_token: GitHub OAuth token
            owner: Repository owner
            repo: Repository name
            issue_number: Issue number
            comment_text: Comment body text
            
        Returns:
            GitHub API response for the created comment
        """
        try:
            async with httpx.AsyncClient() as client:
                # First, verify repo access
                repo_response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}",
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json"
                    },
                    timeout=30.0
                )
                
                if repo_response.status_code != 200:
                    raise HTTPException(
                        status_code=repo_response.status_code,
                        detail=f"Cannot access repository: {repo_response.status_code} {repo_response.text}"
                    )
                
                repo_data = repo_response.json()
                permissions = repo_data.get('permissions', {})
                
                # Check if user has required permissions
                if not (permissions.get('admin') or permissions.get('maintain') or permissions.get('push')):
                    raise HTTPException(
                        status_code=403,
                        detail="Insufficient permissions to comment on this repository"
                    )
                
                # Post the comment
                comment_response = await client.post(
                    f"{self.base_url}/repos/{owner}/{repo}/issues/{issue_number}/comments",
                    json={"body": comment_text},
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json",
                        "Content-Type": "application/json"
                    },
                    timeout=30.0
                )
                
                if comment_response.status_code not in [200, 201]:
                    raise HTTPException(
                        status_code=comment_response.status_code,
                        detail=f"Failed to post comment: {comment_response.status_code} {comment_response.text}"
                    )
                
                return comment_response.json()
                
        except HTTPException:
            raise
        except httpx.TimeoutException:
            raise HTTPException(status_code=504, detail="GitHub API request timed out")
        except Exception as e:
            logger.error(f"Failed to comment on issue #{issue_number} in {owner}/{repo}: {e}")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to comment on issue #{issue_number} in {owner}/{repo}: {str(e)}"
            )
    
    async def fetch_issue_comments(
        self,
        github_access_token: str,
        owner: str,
        repo: str,
        issue_number: int
    ) -> List[Dict]:
        """
        Fetch all comments for a GitHub issue.
        
        Args:
            github_access_token: GitHub OAuth token
            owner: Repository owner
            repo: Repository name
            issue_number: Issue number
            
        Returns:
            List of comment objects from GitHub API
        """
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/repos/{owner}/{repo}/issues/{issue_number}/comments",
                    headers={
                        "Authorization": f"Bearer {github_access_token}",
                        "Accept": "application/vnd.github+json"
                    },
                    timeout=30.0
                )
                
                if response.status_code != 200:
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"Failed to fetch comments: {response.status_code} {response.text}"
                    )
                
                return response.json()
                
        except HTTPException:
            raise
        except httpx.TimeoutException:
            raise HTTPException(status_code=504, detail="GitHub API request timed out")
        except Exception as e:
            logger.error(f"Failed to fetch comments for issue #{issue_number} in {owner}/{repo}: {e}")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to fetch comments for issue #{issue_number} in {owner}/{repo}: {str(e)}"
            )
    
    async def fetch_repo_issues(self, repo_full_name: str, github_access_token: Optional[str] = None, include_prs: bool = True) -> Dict:

        """Fetch issues and PRs from a GitHub repository."""
        try:
            async with httpx.AsyncClient() as client:
                issues_url = f"{self.base_url}/repos/{repo_full_name}/issues"
                params = {"state": "all", "per_page": 100}
                headers = {"Accept": "application/vnd.github+json"}
                if github_access_token:
                    headers["Authorization"] = f"Bearer {github_access_token}"
                
                response = await client.get(issues_url, params=params, headers=headers, timeout=30.0)
                
                if response.status_code != 200:
                    raise HTTPException(
                        status_code=400, 
                        detail=f"Failed to fetch from GitHub: {response.text}"
                    )
                
                items = response.json()
                issues = []
                prs = []
                
                for item in items:
                    is_pr = 'pull_request' in item
                    if is_pr:
                        prs.append(item)
                    else:
                        issues.append(item)
                
                return {"issues": issues, "prs": prs if include_prs else []}
                
        except Exception as e:
            logger.error(f"GitHub fetch error: {e}")
            raise HTTPException(status_code=500, detail=str(e))
    
    async def fetch_user_activity(self, username: str, github_access_token: Optional[str] = None) -> Dict:
        """Fetch a user's GitHub activity (issues and PRs)."""
        try:
            async with httpx.AsyncClient() as client:
                search_url = f"{self.base_url}/search/issues"
                params = {
                    "q": f"author:{username}",
                    "per_page": 100,
                    "sort": "created",
                    "order": "desc"
                }
                headers = {"Accept": "application/vnd.github+json"}
                if github_access_token:
                    headers["Authorization"] = f"Bearer {github_access_token}"
                
                response = await client.get(search_url, params=params, headers=headers, timeout=30.0)
                
                if response.status_code != 200:
                    logger.error(f"GitHub search error: {response.text}")
                    return {"issues": [], "prs": []}
                
                items = response.json().get('items', [])
                issues = []
                prs = []
                
                for item in items:
                    is_pr = 'pull_request' in item
                    if is_pr:
                        prs.append(item)
                    else:
                        issues.append(item)
                
                return {"issues": issues, "prs": prs}
                
        except Exception as e:
            logger.error(f"GitHub user activity fetch error: {e}")
            return {"issues": [], "prs": []}


# Singleton instance
github_service = GitHubService()

File: frontend_new/src/services/githubService.js
/**
 * GitHub API Service
 * Handles direct GitHub API calls for PR management
 */

const GITHUB_API_BASE = "https://api.github.com";

/**
 * Fetch all repositories where the user has maintainer permissions
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string[]} existingRepos - Array of repo full names already added
 * @returns {Promise<Array<{owner: string, repo: string}>>}
 */
export async function fetchMaintainerRepos(
  githubAccessToken,
  existingRepos = [],
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/user/repos?per_page=100&type=owner,collaborator`,
      {
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
        },
      },
    );

    if (!response.ok) {
      throw new Error(
        `GitHub API error: ${response.status} ${response.statusText}`,
      );
    }

    const repos = await response.json();

    const filteredRepos = repos
      .filter(
        (repo) =>
          (repo.permissions?.admin ||
            repo.permissions?.maintain ||
            repo.permissions?.push) &&
          !existingRepos.includes(repo.full_name),
      )
      .map((repo) => ({
        owner: repo.owner.login,
        repo: repo.name,
        fullName: repo.full_name,
        description: repo.description,
        isPrivate: repo.private,
      }));

    return filteredRepos;
  } catch (error) {
    console.error("Failed to fetch maintainer repos:", error);
    throw new Error(`Failed to fetch maintainer repos: ${error.message}`);
  }
}

/**
 * Fetch unmerged pull requests for a repository
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @returns {Promise<Array<{number: number, title: string}>>}
 */
export async function fetchUnmergedPullRequests(
  githubAccessToken,
  owner,
  repo,
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls?state=open`,
      {
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
        },
      },
    );

    if (!response.ok) {
      throw new Error(
        `GitHub API error: ${response.status} ${response.statusText}`,
      );
    }

    const pulls = await response.json();

    const unmergedPRs = pulls
      .filter((pr) => pr.merged_at === null)
      .map((pr) => ({
        number: pr.number,
        title: pr.title,
        user: pr.user.login,
        createdAt: pr.created_at,
        htmlUrl: pr.html_url,
      }));

    return unmergedPRs;
  } catch (error) {
    console.error(`Failed to fetch unmerged PRs for ${owner}/${repo}:`, error);
    throw new Error(
      `Failed to fetch unmerged PRs for ${owner}/${repo}: ${error.message}`,
    );
  }
}

/**
 * Post a comment on a pull request
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} prNumber - Pull request number
 * @param {string} commentText - Comment body text
 * @returns {Promise<Object>} GitHub API response
 */
export async function commentOnPullRequest(
  githubAccessToken,
  owner,
  repo,
  prNumber,
  commentText,
) {
  try {
    // Verify repo access
    const repoResponse = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}`,
      {
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
        },
      },
    );

    if (!repoResponse.ok) {
      throw new Error(
        `Cannot access repository: ${repoResponse.status} ${repoResponse.statusText}`,
      );
    }

    const repoData = await repoResponse.json();

    if (
      !repoData.permissions?.admin &&
      !repoData.permissions?.maintain &&
      !repoData.permissions?.push
    ) {
      throw new Error("Insufficient permissions to comment on this repository");
    }

    // Post comment
    const commentResponse = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/issues/${prNumber}/comments`,
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ body: commentText }),
      },
    );

    if (!commentResponse.ok) {
      throw new Error(
        `Failed to post comment: ${commentResponse.status} ${commentResponse.statusText}`,
      );
    }

    return await commentResponse.json();
  } catch (error) {
    console.error(
      `Failed to comment on PR #${prNumber} in ${owner}/${repo}:`,
      error,
    );
    throw new Error(
      `Failed to comment on PR #${prNumber} in ${owner}/${repo}: ${error.message}`,
    );
  }
}

/**
 * Merge a pull request
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} prNumber - Pull request number
 * @param {string} mergeMethod - Merge method: 'merge', 'squash', or 'rebase'
 * @returns {Promise<Object>} GitHub API response
 */
export async function mergePullRequest(
  githubAccessToken,
  owner,
  repo,
  prNumber,
  mergeMethod = "merge",
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls/${prNumber}/merge`,
      {
        method: "PUT",
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ merge_method: mergeMethod }),
      },
    );

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(
        errorData.message ||
          `Failed to merge PR: ${response.status} ${response.statusText}`,
      );
    }

    return await response.json();
  } catch (error) {
    console.error(
      `Failed to merge PR #${prNumber} in ${owner}/${repo}:`,
      error,
    );
    throw new Error(`Failed to merge PR: ${error.message}`);
  }
}

/**
 * Close a pull request or issue
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} number - Issue/PR number
 * @param {boolean} isPR - Whether this is a pull request
 * @returns {Promise<Object>} GitHub API response
 */
export async function closeIssueOrPR(
  githubAccessToken,
  owner,
  repo,
  number,
  isPR = false,
) {
  try {
    const endpoint = isPR
      ? `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls/${number}`
      : `${GITHUB_API_BASE}/repos/${owner}/${repo}/issues/${number}`;

    const response = await fetch(endpoint, {
      method: "PATCH",
      headers: {
        Authorization: `Bearer ${githubAccessToken}`,
        Accept: "application/vnd.github+json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ state: "closed" }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(
        errorData.message ||
          `Failed to close: ${response.status} ${response.statusText}`,
      );
    }

    return await response.json();
  } catch (error) {
    console.error(`Failed to close #${number} in ${owner}/${repo}:`, error);
    throw new Error(`Failed to close: ${error.message}`);
  }
}

/**
 * Fetch the raw diff for a pull request
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} prNumber - Pull request number
 * @returns {Promise<string>} Raw diff content
 */
export async function fetchPullRequestDiff(
  githubAccessToken,
  owner,
  repo,
  prNumber,
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls/${prNumber}`,
      {
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github.v3.diff",
        },
      },
    );

    if (!response.ok) {
      throw new Error(
        `Failed to fetch diff: ${response.status} ${response.statusText}`,
      );
    }

    return await response.text();
  } catch (error) {
    console.error(
      `Failed to fetch diff for PR #${prNumber} in ${owner}/${repo}:`,
      error,
    );
    throw new Error(`Failed to fetch diff: ${error.message}`);
  }
}

/**
 * Fetch changed files for a pull request
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} prNumber - Pull request number
 * @returns {Promise<Array>} List of changed files with patches
 */
export async function fetchPullRequestFiles(
  githubAccessToken,
  owner,
  repo,
  prNumber,
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls/${prNumber}/files`,
      {
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
        },
      },
    );

    if (!response.ok) {
      throw new Error(
        `Failed to fetch files: ${response.status} ${response.statusText}`,
      );
    }

    return await response.json();
  } catch (error) {
    console.error(
      `Failed to fetch files for PR #${prNumber} in ${owner}/${repo}:`,
      error,
    );
    throw new Error(`Failed to fetch files: ${error.message}`);
  }
}

/**
 * Submit a review on a pull request
 * @param {string} githubAccessToken - GitHub OAuth token
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {number} prNumber - Pull request number
 * @param {string} event - Review event: 'APPROVE', 'REQUEST_CHANGES', or 'COMMENT'
 * @param {string} body - Review comment body
 * @returns {Promise<Object>} GitHub API response
 */
export async function submitPullRequestReview(
  githubAccessToken,
  owner,
  repo,
  prNumber,
  event,
  body = "",
) {
  try {
    const response = await fetch(
      `${GITHUB_API_BASE}/repos/${owner}/${repo}/pulls/${prNumber}/reviews`,
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${githubAccessToken}`,
          Accept: "application/vnd.github+json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          body,
          event, // APPROVE, REQUEST_CHANGES, or COMMENT
        }),
      },
    );

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(
        errorData.message ||
          `Failed to submit review: ${response.status} ${response.statusText}`,
      );
    }

    return await response.json();
  } catch (error) {
    console.error(
      `Failed to submit review for PR #${prNumber} in ${owner}/${repo}:`,
      error,
    );
    throw new Error(`Failed to submit review: ${error.message}`);
  }
}
